tests/smoke/amg.cpp

```diff
@@ -0,0 +1,713 @@
+
+/*
+ *   Copyright 2022 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Test for AMG solver, levels provided by AMGCL.
+ * @file amg.cpp
+ * @author Alberto Scolari (alberto.scolari@huawei.com)
+ * @author Denis Jelovina (denis.jelovina@huawei.com)
+ *
+ *
+ * @date 2022-10-08
+ */
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <type_traits>
+#include <string>
+#include <graphblas.hpp>
+#include <graphblas/algorithms/amg/amg.hpp>
+#include <graphblas/algorithms/amg/system_building_utils.hpp>
+#include <graphblas/utils/Timer.hpp>
+#include <utils/argument_parser.hpp>
+#include <utils/assertions.hpp>
+#include <utils/print_vec_mat.hpp>
+
+// forward declaration for the tracing facility
+template< typename T,
+	class Ring = grb::Semiring<
+		grb::operators::add< T >,
+		grb::operators::mul< T >,
+		grb::identities::zero,
+		grb::identities::one
+	>
+>
+void print_norm( const grb::Vector< T > &r, const char * head, const Ring &ring = Ring() );
+
+#ifdef AMG_PRINT_STEPS
+
+// AMG_PRINT_STEPS requires defining the following symbols
+
+/**
+ * simply prints \p args on a dedicated line.
+ */
+#define DBG_println( args ) std::cout << args << std::endl;
+
+/**
+ * prints \p head and the norm of \p r.
+ */
+#define DBG_print_norm( vec, head ) print_norm( vec, head )
+#endif
+
+//========== MAIN PROBLEM PARAMETERS =========
+// values modifiable via cmd line args: default set as in reference AMG
+constexpr size_t DEF_COARSENING_LEVELS{ 1U };
+constexpr size_t MAX_COARSENING_LEVELS{ 4U };
+constexpr size_t MAX_ITERATIONS_DEF{ 56UL };
+constexpr size_t SMOOTHER_STEPS_DEF{ 1 };
+//============================================
+
+constexpr double MAX_NORM { 4.0e-14 };
+
+using namespace grb;
+using namespace algorithms;
+
+static const char * const TEXT_HIGHLIGHT = "===> ";
+#define thcout ( std::cout << TEXT_HIGHLIGHT )
+#define thcerr ( std::cerr << TEXT_HIGHLIGHT )
+
+/**
+ * Container to store matrices loaded from a file.
+ */
+template< typename T = double >
+struct mat_data {
+	size_t *i_data;
+	size_t *j_data;
+	T *v_data;
+	size_t nz, n, m;
+
+	void resize( size_t innz, size_t inn, size_t inm ){
+		nz=innz;
+		n=inn;
+		m=inm;
+		i_data = new size_t [ nz ];
+		j_data = new size_t [ nz ];
+		v_data = new T [ nz ];
```

Comment:
C vs. C++ (but not critical here since it's a test, feel free to ignore): why not use std::vector in this case? Elsewhere we do use raw pointers, but we couple it with non-trivial allocation mechanisms and/or data ownership hacks